File: src/main/java/com/insano10/gham/PullRequestAnalyser.java
package com.insano10.gham;

import com.insano10.gham.entities.Comment;
import com.insano10.gham.entities.PullRequest;

import java.time.temporal.ChronoUnit;
import java.util.*;

public class PullRequestAnalyser
{

    public Map<String, PullRequestSummary> analyse(List<PullRequest> pullRequests)
    {
        final Map<String, PullRequestSummary> summaries = new HashMap<>();

        for (PullRequest pullRequest : pullRequests)
        {
            final PullRequestSummary summary = getSummary(pullRequest.getOwner(), summaries);

            final long minsOpen = pullRequest.getCreated().until(pullRequest.getClosed(), ChronoUnit.MINUTES);
            summary.pullRequestRaised(minsOpen);

            Set<String> commentersSeen = new HashSet<>();
            for (Comment comment : pullRequest.getComments())
            {
                if (!commentersSeen.contains(comment.getOwner()))
                {
                    final PullRequestSummary commenterSummary = getSummary(comment.getOwner(), summaries);

                    final long minsTillFirstComment = pullRequest.getCreated().until(comment.getCreated(), ChronoUnit.MINUTES);

                    commenterSummary.pullRequestCommentedOn(minsTillFirstComment);
                    commentersSeen.add(comment.getOwner());
                }
            }
        }

        return summaries;
    }

    private PullRequestSummary getSummary(String user, Map<String, PullRequestSummary> summaries)
    {
        return summaries.computeIfAbsent(user, PullRequestSummary::new);
    }
}

File: src/main/scala/com/insano10/github/PullRequestAnalyser.scala
package com.insano10.github

import java.time.temporal.ChronoUnit

import com.insano10.github.entities.{PullRequestSummary, PullRequest}

class PullRequestAnalyser {

  def analyse(pullRequests: List[PullRequest]): List[PullRequestSummary] = {

    var summaries: Map[String, PullRequestSummary] = Map()

    pullRequests.foreach(pullRequest => {

      val summary = getSummary(pullRequest.owner, summaries)
      val minsOpen: Long = pullRequest.created.until(pullRequest.closed, ChronoUnit.MINUTES)

      summary.pullRequestRaised(minsOpen)
      var commentersSeen = Set[String]()

      for (comment <- pullRequest.comments) {

        if (!commentersSeen.contains(comment.owner) && !(comment.owner == pullRequest.owner)) {

          val commenterSummary = getSummary(comment.owner, summaries)
          val minsTillFirstComment: Long = pullRequest.created.until(comment.created, ChronoUnit.MINUTES)
          commenterSummary.pullRequestCommentedOn(minsTillFirstComment)
          commentersSeen += comment.owner
          summaries = summaries.updated(comment.owner, commenterSummary)
        }
      }
      summaries = summaries.updated(pullRequest.owner, summary)
    })

    summaries.values.toList
  }

  private def getSummary(user: String, summaries: Map[String, PullRequestSummary]) = {

    summaries.getOrElse(user, new PullRequestSummary(user))
  }
}
